numeric stage references can't be resolved

**Actual behavior**
When referencing stages in numeric format

```dockerfile
# integration/dockerfiles/Dockerfile_test_copy_chown_nonexisting_user
FROM debian:10.13
RUN echo "hey" > /tmp/foo

FROM debian:10.13
RUN groupadd --gid 5000 testgroup
COPY --from=0 --chown=1001:1001 /tmp/foo /tmp/baz
# with existing group
COPY --from=0 --chown=1001:testgroup /tmp/foo /tmp/baz
```
when combined with `--skip-unused-stages`, kaniko comes to a crieking halt.

https://github.com/mzihlmann/kaniko/actions/runs/16092608173/job/45411287599
```
         images.go:522: Going to build image with kaniko: localhost:5000/kaniko-dockerfile_test_copy_chown_nonexisting_user, flags: [-c /workspace] 
        integration_test.go:644: Error building image: Failed to build image localhost:5000/kaniko-dockerfile_test_copy_chown_nonexisting_user with kaniko command "[docker run --net=host -e BENCHMARK_FILE=false -v /home/runner/work/kaniko/kaniko/integration:/workspace -v /tmp/2160908095:/kaniko/benchmarks -v /home/runner/.docker/config.json:/root/.docker/config.json -e DOCKER_CONFIG=/root/.docker executor-image -f /workspace/dockerfiles/Dockerfile_test_copy_chown_nonexisting_user -d localhost:5000/kaniko-dockerfile_test_copy_chown_nonexisting_user --force -c /workspace]": exit status 1
            INFO[0000] Using dockerignore file: /workspace/.dockerignore 
            INFO[0000] Retrieving image manifest 0                  
            error building image: could not parse reference: 0
```



**Expected behavior**
Kaniko should work with numeric stages as docker does.

--- a/integration/images.go
+++ b/integration/images.go
@@ -93,6 +93,8 @@ var additionalKanikoFlagsMap = map[string][]string{
 	"Dockerfile_test_cache_copy":             {"--cache-copy-layers=true"},
 	"Dockerfile_test_cache_copy_oci":         {"--cache-copy-layers=true"},
 	"Dockerfile_test_issue_add":              {"--cache-copy-layers=true"},
+	// skip-unused-stages
+	"Dockerfile_test_copy_chown_nonexisting_user": {"--skip-unused-stages"},
 }
 
 // Arguments to diffoci when comparing dockerfiles

--- a/pkg/dockerfile/dockerfile.go
+++ b/pkg/dockerfile/dockerfile.go
@@ -349,48 +349,53 @@ func unifyArgs(metaArgs []instructions.ArgCommand, buildArgs []string) []string 
 
 // skipUnusedStages returns the list of used stages without the unnecessaries ones
 func skipUnusedStages(stages []instructions.Stage, lastStageIndex *int, target string) []instructions.Stage {
-	stagesDependencies := make(map[string]bool)
-	var onlyUsedStages []instructions.Stage
-	idx := *lastStageIndex
+	stageByName := make(map[string]int)
 
-	lastStageBaseName := stages[idx].BaseName
+	for idx, s := range stages {
+		if s.Name != "" {
+			stageByName[s.Name] = idx
+		}
+	}
 
-	for i := idx; i >= 0; i-- {
+	stagesDependencies := make([]bool, len(stages))
+	stagesDependencies[*lastStageIndex] = true
+
+	for i := *lastStageIndex; i >= 0; i-- {
+		if !stagesDependencies[i] {
+			continue
+		}
 		s := stages[i]
-		if (s.Name != "" && stagesDependencies[s.Name]) || s.Name == lastStageBaseName || i == idx {
-			for _, c := range s.Commands {
-				switch cmd := c.(type) {
-				case *instructions.CopyCommand:
-					stageName := cmd.From
-					if copyFromIndex, err := strconv.Atoi(stageName); err == nil {
-						stageName = stages[copyFromIndex].Name
-					}
-					if !stagesDependencies[stageName] {
-						stagesDependencies[stageName] = true
+		if BaseIndex, ok := stageByName[s.BaseName]; ok {
+			// There can be references that appear as non-existing stages
+			// ie. `FROM debian AS base` would try refer to `debian` as stage
+			// before falling back to `debian` as a docker image.
+			stagesDependencies[BaseIndex] = true
+		}
+		for _, c := range s.Commands {
+			switch cmd := c.(type) {
+			case *instructions.CopyCommand:
+				if copyFromIndex, err := strconv.Atoi(cmd.From); err == nil {
+					// numeric reference `COPY --from=0`
+					stagesDependencies[copyFromIndex] = true
+				} else {
+					// named reference `COPY --from=base`
+					if copyFromIndex, ok := stageByName[cmd.From]; ok {
+						// There can be references that appear as non-existing stages
+						// ie. `COPY --from=debian` would try refer to `debian` as stage
+						// before falling back to `debian` as a docker image.
+						stagesDependencies[copyFromIndex] = true
 					}
 				}
 			}
-			if i != idx {
-				stagesDependencies[s.BaseName] = true
-			}
 		}
-	}
-	dependenciesLen := len(stagesDependencies)
-	if dependenciesLen > 0 {
-		for i := 0; i < idx; i++ {
-			if stages[i].Name == "" {
-				continue
-			}
-			s := stages[i]
-			if stagesDependencies[s.Name] || s.Name == lastStageBaseName {
-				onlyUsedStages = append(onlyUsedStages, s)
-			}
-		}
-	}
-	onlyUsedStages = append(onlyUsedStages, stages[idx])
-	if idx > len(onlyUsedStages)-1 {
-		*lastStageIndex = len(onlyUsedStages) - 1
 	}
 
+	var onlyUsedStages []instructions.Stage
+	for i := 0; i < *lastStageIndex+1; i++ {
+		if stagesDependencies[i] {
+			onlyUsedStages = append(onlyUsedStages, stages[i])
+		}
+	}
+	*lastStageIndex = len(onlyUsedStages) - 1
 	return onlyUsedStages
 }

